Adding filestore with DB backend

nicolaferraro opened the pull request:
First PR on syndesis :smile: 
This fixes #735.

It adds a Filestore that can be used to store binary artifacts such as JARs.
The interface allows to do operation on filesystem-like paths and can be adapted to store artifacts in whatever storage we want in the future.

Current implementation uses a DB table called `filestore` inside the postgres database with 2 columns (`path` and `data`).

All the required file operation should be already implemented (write, load, move, delete and temp file support). Others can be added as needed.

nicolaferraro commented:
@KurtStam do I need to fix *all* codacy "errors" or it's there as suggestion?

rhuss commented:
We all love  codacy ;-) It's ok to fix only the reasonable findings, and just ignore the stupid ones (we laready have quite struct PMD and findbugs checks, so codacy adds some sugar on top of it).

KurtStam commented on filestore/src/main/java/io/syndesis/filestore/FileStoreException.java:
```diff
@@ -0,0 +1,30 @@
+    public FileStoreException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}
```
Can you add a launderThrowable method and use it? Like https://github.com/syndesisio/syndesis-rest/blob/master/core/src/main/java/io/syndesis/core/SyndesisServerException.java#L43-L55

nicolaferraro replied:
I can add a launderThrowable but the filestore exception is always used to add context information (e.g. the path being written) when a low level exception occurs. Since DBI exceptions (the low level exceptions) are always runtime exceptions, the message would be discarded, the runtime exception rethrown as is.. and this exception would not be used..

KurtStam replied:
Yes ignore my comment, you actually want it to be a runtime exception.

KurtStam commented on filestore/src/main/java/io/syndesis/filestore/impl/SqlFileStore.java:
```diff
@@ -0,0 +1,188 @@
+public class SqlFileStore implements FileStore {
+
+    enum DatabaseKind {
+        PostgreSQL, H2
```
Most of our DB testing is done on Derby, can we add support for that too? Not sure we will use H2, we may want to drop that to simplify maintenance.

nicolaferraro commented:
@KurtStam I've implemented some changes, including testing on derby and `launderThrowable` where best applicable.

Unfortunately I discovered I was too optimistic on thinking that BLOB support was handled correctly by JDBI, but after testing with Derby and with a external Postgres instance, I've added some custom code to handle the specific details of each database.

It would be good if I can replicate in CI the tests that I did manually on postgres, because there's a lot of difference between the way Postgres, H2 and Derby manage BLOBs. Are such tests supported by current CI?

KurtStam commented:
Thanks @nicolaferraro, unfortunately only our e2e tests run with Postgres at the moment.

rhuss commented:
@nicolaferraro is this an intrusive PR ? Just asking because we are still on the TP2 train. BTW, we wanted to introduce feature toggle to switch on/off things while we are going. this is just starting, though.

nicolaferraro commented:
I've added a `filestore.enabled` flag for this, and it defaults to `false`, so the filestore is not created (call it a basic version of feature toggle).

Just checked on syndesis to be sure.

rhuss commented:
@nicolaferraro cool, thanks. was just asking ;-)

Wonder though whether we should group feature toggles, so to have a top-leve `toggles` element with sub-elements like `filestore`. So excactly the other way round. So its easier to see what feature s toggles are available.
